src/MergeContext.js

```diff
@@ -454,6 +454,214 @@ class BranchPosition
+    {
+        if (!this._author)
+            return this._defaultAuthor;
+        return {name: this._author.name, email: this._author.email, date: this._defaultAuthor.date};
+    }
+
+    // returns the position of the first non-ASCII_printable character (or -1)
+    _invalidCharacterPosition(str) {
+        const prohibitedCharacters = /[^\u{20}-\u{7e}]/u;
+        const match = prohibitedCharacters.exec(str);
+        return match ? match.index : -1;
+    }
+
+    // removes leading empty lines and trims the end
+    _trim(str) {
+        // cannot just use trim() to preserve the first line indentation (if any).
+        return str.replace(/^\s*\n+/g, '').trimEnd();
+    }
+
+    // basic checks for an unparsed message
+    _checkRaw(message) {
+        // CRs in CRLF sequences already removed
+        // other CRs are not treated specially (and are banned)
+        const trimmedMessage = this._trim(message);
+        const lines = trimmedMessage.split('\n');
+        for (let i = 0; i < lines.length; ++i) {
+            const line = lines[i];
+            const invalidPosition = this._invalidCharacterPosition(line);
+
+            if (invalidPosition !== -1)
+                throw new Error(`bad character at line ${i}, offset ${invalidPosition}`);
+
+            if (line.length > 72)
+                throw new Error(`too long line '${line}'`);
+        }
+    }
+
+    // parses attributes of the raw PR description
+    _parse(prDescription) {
+        const prDescriptionWithoutHeader = this._extractHeader(prDescription);
+
+        try {
+            const prDescriptionWithoutHeaderAndTrailer = this._extractTrailer(prDescriptionWithoutHeader);
+            this._parseBody(prDescriptionWithoutHeaderAndTrailer);
+        } catch (e) {
+            // TODO: supply debugString() prefix
+            Log.Logger.info("assuming no trailer: " + e.message);
+            this._parseBody(prDescriptionWithoutHeader);
+        }
+
+        assert(this._body !== undefined && this._body !== null);
+    }
+
+    // author is a {name, value}
+    _parseAuthor(author) {
+        const cred = author.value.match(/^ ([\w][^@<>,]*) <(\S+@\S+\.\S+)>$/);
+        if (!cred)
+            throw new Error(`invalid '${author.name}' value format`);
+
+        if (cred[0].includes(','))
+            throw new Error(`invalid '${author.name}' value: commas are not allowed`);
+
+        return {name: cred[1].trim(), email: cred[2].trim()};
+    }
+
+    _extractHeader(prDescriptionRaw) {
+        const prDescription = this._trim(prDescriptionRaw);
+        if (prDescription.startsWith('Authored-by:')) {
+            let tokenizer = new FieldsTokenizer(prDescription);
+            const authorField = tokenizer.nextField();
+            this._author = this._parseAuthor(authorField);
+            if (!tokenizer.atEnd())
+                throw new Error(`header must have nothing but a single 'Authored-by' attribute`);
+            return this._trim(tokenizer.remaining());
```

Comment:
This trimming should not be our responsibility. Placing this trimming code here implies that the caller can count on us trimming the value for them, but we do not want to create that kind of dependency. Each parser should trim its raw input.

```suggestion
            return tokenizer.remaining();
```